File: sdk/go/crawl.go
package contextos

import (
	"context"
	"encoding/json"
	"fmt"
)

// CrawlRequest is the request to crawl a URL.
type CrawlRequest struct {
	URL      string            `json:"url"`
	Depth    int               `json:"depth,omitempty"`
	MaxPages int               `json:"max_pages,omitempty"`
	Extract  bool              `json:"extract,omitempty"`
	Options  map[string]string `json:"options,omitempty"`
}

// CrawlResult represents a crawl result.
type CrawlResult struct {
	ID       string            `json:"id"`
	URL      string            `json:"url"`
	Title    string            `json:"title"`
	Content  string            `json:"content"`
	Links    []string          `json:"links,omitempty"`
	Metadata map[string]string `json:"metadata,omitempty"`
}

// CrawlResponse is the response from crawl.
type CrawlResponse struct {
	Result  CrawlResult `json:"result"`
	Status  string      `json:"status"`
	Message string      `json:"message,omitempty"`
}

// CrawlService handles crawl operations.
type CrawlService struct {
	client *Client
}

// Crawl crawls a URL.
func (s *CrawlService) Crawl(ctx context.Context, req *CrawlRequest) (*CrawlResponse, error) {
	var result CrawlResponse
	body, err := s.client.doRequest(ctx, "POST", "/api/v1/crawl", req)
	if err != nil {
		return nil, err
	}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}
	return &result, nil
}

// Map crawls and maps a website.
func (s *CrawlService) Map(ctx context.Context, url string, maxPages int) (*CrawlResponse, error) {
	req := &CrawlRequest{
		URL:      url,
		MaxPages: maxPages,
	}
	return s.Crawl(ctx, req)
}

// Extract extracts content from a URL.
func (s *CrawlService) Extract(ctx context.Context, url string) (*CrawlResponse, error) {
	req := &CrawlRequest{
		URL:     url,
		Extract: true,
	}
	return s.Crawl(ctx, req)
}

File: sdk/rust/src/crawl.rs
use serde::{Deserialize, Serialize};

use crate::error::Result;

/// Crawl request.
#[derive(Debug, Serialize)]
pub struct CrawlRequest {
    pub url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub depth: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_pages: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extract: Option<bool>,
}

/// Crawl result.
#[derive(Debug, Deserialize)]
pub struct CrawlResult {
    pub id: String,
    pub url: String,
    pub title: String,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub links: Option<Vec<String>>,
}

/// Crawl response.
#[derive(Debug, Deserialize)]
pub struct CrawlResponse {
    pub result: CrawlResult,
    pub status: String,
}

/// Crawl service.
pub struct CrawlService {
    base_url: String,
    api_key: String,
}

impl CrawlService {
    pub(crate) fn new(base_url: String, api_key: String) -> Self {
        CrawlService { base_url, api_key }
    }

    /// Crawl a URL.
    pub async fn crawl(&self, url: &str) -> Result<CrawlResponse> {
        let req = CrawlRequest {
            url: url.to_string(),
            depth: None,
            max_pages: None,
            extract: None,
        };

        let url = format!("{}/api/v1/crawl", self.base_url);
        let resp = reqwest::Client::new()
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .json(&req)
            .send()
            .await
            .map_err(|e| crate::error::Error::Request(e.to_string()))?;

        let result: CrawlResponse = resp
            .json()
            .await
            .map_err(|e| crate::error::Error::Serialization(e.to_string()))?;

        Ok(result)
    }

    /// Extract content from a URL.
    pub async fn extract(&self, url: &str) -> Result<CrawlResponse> {
        let req = CrawlRequest {
            url: url.to_string(),
            depth: None,
            max_pages: None,
            extract: Some(true),
        };

        let url = format!("{}/api/v1/crawl", self.base_url);
        let resp = reqwest::Client::new()
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .json(&req)
            .send()
            .await
            .map_err(|e| crate::error::Error::Request(e.to_string()))?;

        let result: CrawlResponse = resp
            .json()
            .await
            .map_err(|e| crate::error::Error::Serialization(e.to_string()))?;

        Ok(result)
    }
}
